shared/service_registry.go

```diff
@@ -0,0 +1,71 @@
+package shared
+
+import (
+	"fmt"
+	"log"
+	"reflect"
+)
+
+// ServiceRegistry provides a useful pattern for managing services attached to a struct.
+// This can be used for managing routines that are part of items such as a
+// beacon chain node, or a sharding client. That is, it allows for ease of dependency
+// management and ensures services dependent on others use the same references in memory.
+//
+// A beacon chain node, for example, has various services including p2p, proposers, attesters,
+// and more. Proposers depend on p2p, so upon registering a proposer service, a p2p service
+// can be fetched from the service registry and passed into the p2p constructor.
+// See beacon-chain/node/node.go for examples.
```

Comment:
I think we should add service_registery_example_test.go if we want to recommend examples.

Additionally, this comment block has a lot of information about the outside world. Can we remove ideas of beacon chain or sharding or anything specific for this generic service registry?